index.d.ts

```diff
@@ -75,48 +75,69 @@ type EncodingOption =
 type DefaultEncodingOption = 'utf8';
 type BufferEncodingOption = 'buffer';
 
-type TupleItem<
-	Tuple extends readonly unknown[],
-	Index extends string,
-> = Tuple[Index extends keyof Tuple ? Index : number];
-
 // Whether `result.stdout|stderr|all` is `undefined`, excluding the `buffer` option
 type IgnoresStreamResult<
 	StreamIndex extends string,
 	OptionsType extends Options = Options,
 	// `result.stdin` is always `undefined`
 > = StreamIndex extends '0' ? true
 	// When using `stdout|stderr: 'inherit'`, or `'ignore'`, etc. , `result.std*` is `undefined`
-	: OptionsType[TupleItem<StdioOptionNames, StreamIndex>] extends NoOutputStdioOption ? true
-		// Same but with `stdio: 'ignore'`
-		: OptionsType['stdio'] extends NoOutputStdioOption ? true
-			// Same but with `stdio: ['ignore', 'ignore', 'ignore', ...]`
-			: OptionsType['stdio'] extends StdioOptionsArray
-				? TupleItem<OptionsType['stdio'], StreamIndex> extends StdioOptionsArray[number]
-					? IgnoresStdioResult<TupleItem<OptionsType['stdio'], StreamIndex>>
-					: false
-				: false;
+	: IgnoresNormalPropertyResult<StreamIndex, OptionsType> extends true ? true
+		// Otherwise
+		: IgnoresStdioPropertyResult<StreamIndex, OptionsType['stdio']>;
+
+type IgnoresNormalPropertyResult<
+	StreamIndex extends string,
+	OptionsType extends Options = Options,
+> = StreamIndex extends keyof StdioOptionNames
+	? StdioOptionNames[StreamIndex] extends keyof OptionsType
+		? OptionsType[StdioOptionNames[StreamIndex]] extends NoOutputStdioOption
+			? true
+			: false
+		: false
+	: false;
 
 type StdioOptionNames = ['stdin', 'stdout', 'stderr'];
 
+type IgnoresStdioPropertyResult<
+	StreamIndex extends string,
+	StdioOptionType extends StdioOptions | undefined,
+	// Same but with `stdio: 'ignore'`
+> = StdioOptionType extends NoOutputStdioOption ? true
+	// Same but with `stdio: ['ignore', 'ignore', 'ignore', ...]`
+	: StdioOptionType extends StdioOptionsArray
+		? StreamIndex extends keyof StdioOptionType
+			? StdioOptionType[StreamIndex] extends StdioOption
+				? IgnoresStdioResult<StdioOptionType[StreamIndex]>
+				: false
+			: false
+		: false;
+
 // Whether `result.stdio[*]` is `undefined`
-type IgnoresStdioResult<
-	StdioOption extends StdioOptionsArray[number],
-> = StdioOption extends NoOutputStdioOption
-	? true
+type IgnoresStdioResult<StdioOptionType extends StdioOption> =
+	StdioOptionType extends NoOutputStdioOption ? true
 	// `result.stdio[3+]` is `undefined` when it is an input stream
-	: StdioOption extends StdinOption
-		? StdioOption extends StdoutStderrOption
-			? false
-			: true
-		: false;
+		: StdioOptionType extends StdinOption
+			? StdioOptionType extends StdoutStderrOption
+				? false
+				: true
+			: false;
 
 // Whether `result.stdout|stderr|all` is `undefined`
 type IgnoresStreamOutput<
 	StreamIndex extends string,
 	OptionsType extends Options = Options,
-> = OptionsType extends {buffer: false} ? true
```

Comment:
Small refactoring to prefer `ObjectType[Property] extends boolean` over `ObjectType extends {Property: boolean}` since the first one correctly distributes union types ("distributive conditional types"), but not the second one.